components/eam/src/physics/cam/zm_conv_intr.F90

```diff
@@ -402,23 +1044,217 @@ subroutine zm_conv_tend(pblh    ,mcon    ,cme     , &
+       endif
+     enddo
+
+     Qm(:,:) = 0.0_r8
+     Qmq(:,:) = 0.0_r8
+     Qmu(:,:) = 0.0_r8
+     Qmv(:,:) = 0.0_r8
+
+     do i = 1,ncol
+
+      Qm_int_end(i) = 0._r8
+      Qmq_int_end(i) = 0._r8
+      Pa_int_end(i) = 0._r8
+
+      Q_dis(i) = 0._r8
+      Qq_dis(i) = 0._r8
+
+      if( (state%pint(i,pver+1)-state%pmid(i,jctop(i))) >= 70000._r8 ) then
+      if(abs(du600(i)).ge.3._r8.and.abs(du600(i)).lt.200.and.Qs_zmconv(i).gt.0._r8) then
+
+       do k = jctop(i),pver
+
+           top = state%pint(i,pver+1) - state%pmid(i,k)
+           bottom = state%pint(i,pver+1) - state%pmid(i,jctop(i))
+
+           Qm(i,k) = -1._r8 * Qs_zmconv(i) * alpha2 * sin(2.0_r8*pi*(top/bottom))
+           Qmq(i,k) = -1._r8 * Qv_zmconv(i) * alpha2 * sin(2.0_r8*pi*(top/bottom))
+           Qmq(i,k) = Qm(i,k)/2500000._r8/4._r8
+
+
+           !if( cos(pi*(top/bottom))>0 ) then 
+              Qmu(i,k) = alphau * (cos(pi*(top/bottom)))
+           !else 
+           !   Qmu(i,k) = alphau * (cos(pi*(top/bottom)))*.25_r8
+           !end if
+
+           Qmv(i,k) = alphav * (cos(pi*(top/bottom)))
+
+           Qm_int_end(i) = Qm_int_end(i) + Qm(i,k) * (state%pdel(i,k)/gravit)
+           Qm_int_end(i) = Qm_int_end(i) + (2._r8*Qmu(i,k)*ztodt*state%u(i,k)+ &
+                                            Qmu(i,k)*Qmu(i,k)*ztodt*ztodt)/2._r8 * (state%pdel(i,k)/gravit)/ztodt
+           Qm_int_end(i) = Qm_int_end(i) + (2._r8*Qmv(i,k)*ztodt*state%v(i,k)+ &
+                                            Qmv(i,k)*Qmv(i,k)*ztodt*ztodt)/2._r8 * (state%pdel(i,k)/gravit)/ztodt
+           Qmq_int_end(i) = Qmq_int_end(i) + Qmq(i,k) * (state%pdel(i,k)/gravit)
+           Pa_int_end(i) = Pa_int_end(i) + state%pdel(i,k)
+        end do
+
+        Q_dis(i) = Qm_int_end(i) / Pa_int_end(i)
+        Qq_dis(i) = Qmq_int_end(i) / Pa_int_end(i)
+
+       endif
+       endif
+     enddo
+
+     MCSP_DT = 0._r8
+     MCSP_freq = 0._r8
+
+     do i = 1,ncol
+       do k = jctop(i),pver
+         Qcq_adjust = ptend_loc%q(i,k,1) - Qq_dis(i) * gravit
+
+         ptend_loc%s(i,k) = ptend_loc%s(i,k) - Q_dis(i)*gravit ! energy fixer
+
+         MCSP_DT(i,k) = -Q_dis(i)*gravit+Qm(i,k)
+         if(abs(Qm(i,k)).gt.0._r8 .and. abs(Qmu(i,k)).gt.0._r8) MCSP_freq(i) = 1._r8
+
+         if (doslop_heat) ptend_loc%s(i,k) = ptend_loc%s(i,k) + Qm(i,k)
+         if (doslop_moisture) ptend_loc%q(i,k,1) = Qcq_adjust + Qmq(i,k)
+         if (doslop_uwind) ptend_loc%u(i,k) = Qmu(i,k)
+         if (doslop_vwind) ptend_loc%v(i,k) = Qmv(i,k)
+       enddo
+     enddo
+
+   endif
+!   
+! End the MCSP parameterization here 
+!   
+   
+
+  MCSP_DT(:ncol,:pver) = MCSP_DT(:ncol,:pver)/cpair
+   call outfld('MCSP_DT    ',MCSP_DT           ,pcols   ,lchnk   )
+   call outfld('MCSP_freq    ',MCSP_freq           ,pcols   ,lchnk   )
+   call outfld('MCSP_DU    ',ptend_loc%u*86400._r8           ,pcols   ,lchnk   )
+   call outfld('MCSP_DV    ',ptend_loc%v*86400._r8           ,pcols   ,lchnk   )
```

Comment:
Just for reference when creating the PR for E3SM master.

The above two calls of outfld need to be conditioned as follows,
```
      if (doslop_uwind) call outfld('MCSP_DU    ',ptend_loc%u*86400._r8           ,pcols   ,lchnk   )
      if (doslop_vwind) call outfld('MCSP_DV    ',ptend_loc%v*86400._r8           ,pcols   ,lchnk   )
```
Otherwise it would attempt to access unallocated ptend_loc%u and ptend_loc%v, which would fail DEBUG run. Non-debug run is inconsequential.

Reply:
Thank you, Wuyin!